hack/generator/pkg/codegen/pipeline_convert_allof_and_oneof_to_objects.go

```diff
@@ -0,0 +1,576 @@
+		func(ctx context.Context, defs astmodel.Types) (astmodel.Types, error) {
+			visitor := astmodel.MakeTypeVisitor()
+
+			// the context here is whether we are selecting spec or status fields
+			visitor.VisitAllOfType = func(this *astmodel.TypeVisitor, it astmodel.AllOfType, ctx interface{}) (astmodel.Type, error) {
+				synth := synthesizer{
+					specOrStatus: ctx.(specOrStatus),
+					defs:         defs,
+					idFactory:    idFactory,
+				}
+
+				object, err := synth.allOfObject(it)
+				if err != nil {
+					return nil, err
+				}
+
+				// we might end up with something that requires re-visiting
+				// e.g. AllOf can turn into a OneOf that we then need to visit
+				return this.Visit(object, ctx)
+			}
+
+			originalVisitOneOf := visitor.VisitOneOfType
+			visitor.VisitOneOfType = func(this *astmodel.TypeVisitor, it astmodel.OneOfType, ctx interface{}) (astmodel.Type, error) {
+				// process children first so that allOfs are resolved
+				result, err := originalVisitOneOf(this, it, ctx)
+				if err != nil {
+					return nil, err
+				}
+
+				if resultOneOf, ok := result.(astmodel.OneOfType); ok {
+					synth := synthesizer{
+						specOrStatus: ctx.(specOrStatus),
+						defs:         defs,
+						idFactory:    idFactory,
+					}
+
+					object, err := synth.oneOfObject(resultOneOf)
+					if err != nil {
+						return nil, err
+					}
+
+					result = object
+				}
+
+				// we might end up with something that requires re-visiting
+				return this.Visit(result, ctx)
+			}
+
+			visitor.VisitResourceType = func(this *astmodel.TypeVisitor, it *astmodel.ResourceType, ctx interface{}) (astmodel.Type, error) {
+				spec, err := this.Visit(it.SpecType(), chooseSpec)
+				if err != nil {
+					return nil, err
+				}
+
+				status, err := this.Visit(it.StatusType(), chooseStatus)
+				if err != nil {
+					return nil, err
+				}
+
+				return it.WithSpec(spec).WithStatus(status), nil
+			}
+
+			result := make(astmodel.Types)
+
+			for _, def := range defs {
+				resourceUpdater := chooseSpec
+				// TODO: we need flags
+				if strings.HasSuffix(def.Name().Name(), "_Status") {
+					resourceUpdater = chooseStatus
+				}
+
+				transformed, err := visitor.VisitDefinition(def, resourceUpdater)
+				if err != nil {
+					return nil, errors.Wrapf(err, "error processing type %v", def.Name())
+				}
+
+				result.Add(*transformed)
+			}
+
+			return result, nil
+		})
+}
+
+type synthesizer struct {
+	specOrStatus specOrStatus
+	idFactory    astmodel.IdentifierFactory
+	defs         astmodel.Types
+}
+
+func (s synthesizer) oneOfObject(oneOf astmodel.OneOfType) (astmodel.Type, error) {
+	// If there's more than one option, synthesize a type.
+	// Note that this is required because Kubernetes CRDs do not support OneOf the same way
+	// OpenAPI does, see https://github.com/Azure/k8s-infra/issues/71
+	var properties []*astmodel.PropertyDefinition
+
+	propertyDescription := "mutually exclusive with all other properties"
```

Comment:
Since this is for human consumption, capitalize please. #nit #trivial
```suggestion
	propertyDescription := "Mutually exclusive with all other properties"
```